subcommand/server-acl-init/command.go

```diff
@@ -739,11 +841,85 @@ func (c *Command) configureConnectInject(logger hclog.Logger, consulClient *api.
 		BindName:    "${serviceaccount.name}",
 		Selector:    c.flagBindingRuleSelector,
 	}
-	return c.untilSucceeds(fmt.Sprintf("creating acl binding rule for %s", authMethodTmpl.Name),
-		func() error {
-			_, _, err := consulClient.ACL().BindingRuleCreate(&abr, nil)
+
+	// Add a namespace if appropriate
+	// If namespaces and mirroring are enabled, this is not necessary because
+	// the binding rule will fall back to being created in the Consul `default`
+	// namespace automatically, as is necessary for mirroring.
+	if c.flagEnableNamespaces && !c.flagEnableSyncK8SNSMirroring {
+		abr.Namespace = c.flagConsulSyncDestinationNamespace
+	}
+
+	// If the binding rule already exists and namespaces are enabled, update it
+	if len(existingRules) > 0 && c.flagEnableNamespaces {
+		// Find the policy that matches our name and description
+		// and that's the ID we need
+		for _, existingRule := range existingRules {
+			if existingRule.BindName == abr.BindName && existingRule.Description == abr.Description {
+				abr.ID = existingRule.ID
+			}
+		}
+
+		// This will only happen if there are existing policies
+		// for this auth method, but none that match the binding
+		// rule set up here in the bootstrap method.
+		if abr.ID == "" {
+			return errors.New("Unable to find a matching ACL binding rule to update")
+		}
+
+		err = c.untilSucceeds(fmt.Sprintf("updating acl binding rule for %s", authMethodTmpl.Name),
```

Comment:
This doesn't work if upgrading from oss to enterprise, we get this error:
```
2020-02-05T01:13:18.160Z [ERROR] Failure: updating acl binding rule for consul-k8s-auth-method: err="Unexpected response code: 500 (Changing an ACL binding rule's namespace is not permitted)"
```

Reply:
Fixed.